rfcs/90-re-support-initialization-via-connection-string.md

```diff
@@ -0,0 +1,154 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-06-17
+deprecates:
+  - 13-remove-config-string.md
+---
+
+# GSP-90: Re-support Initialization Via Connection String
+
+## Background
+
+We had config strings like `s3://bucket_name?access_key=xxxx&secret_key=xxxx` before. This is introduced in [GSP-3] but deprecated by [GSP-13].
+
+There are two reasons why config strings are removed, as mentioned in [GSP-13]:
+
+1. Usability: If an application uses [go-storage]'s config string, then it either exposes the format of config string to end user directly (additional burden for end users) or it has to manipulate its config into our format (additional burden for application developers). And if the application wants to combine its other config into a config string, this is also inconvenient.
+2. Implementation: We didn't find a good way to parse pairs from strings easily and safely.
+
+So we let the users of [go-storage] construct pairs directly using `WithXxx()`. 
+
+But passing string config is indeed more convenient, if not much. And actually parsing string into pairs is not that difficult. We can use a registry mechanism like [GSP-48]'s implementation.
+
+## Proposal
+
+So I propose to support service init from connection string (We rename config string to connection string since this name is more common):
+
+We add the following APIs:
+```go
+func NewServicerFromString(connStr string) (types.Servicer, error) {}
+func NewStoragerFromString(connStr string) (types.Storager, error) {}
+```
+
+### Format
+
+The format of the connection string is (optional parts marked by squared brackets):
+
+`<type>://[<name>][<work_dir>][?key1=value1&...&keyN=valueN]`
+
+- name: storage name, e.g., bucket name. MUST NOT contain /
+- work_dir: For object storage, it is prefix; for fs, it is directory path. MUST start with / for every storage services.
+- For the `key=value` pairs, if `=value` is missing, we just ignore the pair. But `key=` means a pair with a blank value.
+- If their are multiple pairs with the same key, the first one will be picked.
```

Comment:
Is it necessary to give the naming conventions for key? Is the key case sensitive or must match the definitions exactly?

Reply:
In current implementation, it is the same as the name in `toml` file

Reply:
I can specify this in the proposal.